sys/sleepq.c

```diff
@@ -294,17 +308,44 @@ bool sleepq_broadcast(void *wchan) {
   return true;
 }
 
-bool sleepq_abort(thread_t *td) {
+// TODO rename this (and possibly `sleepq_abort`)
+static bool sleepq_abort_reason(thread_t *td, sq_wakeup_t reason) {
   sleepq_chain_t *sc = sc_acquire(td->td_wchan);
   sleepq_t *sq = sq_lookup(sc, td->td_wchan);
   bool aborted = false;
 
   if (sq != NULL) {
-    aborted = sq_wakeup(td, sc, sq, SQ_ABORT);
+    aborted = sq_wakeup(td, sc, sq, reason);
     sq_release(sq);
   }
   sc_release(sc);
 
+  // TODO maybe move this out to sleepq_abort?
```

Comment:
Could you elaborate? It's not obvious at first sight.

Reply:
`sq_timeout` is called during `callout_process` with interrupts disabled so `sched_maybe_preempt` will immediately return. It seems a bit weird to even call it there when the callout is going to end right after this call.

Reply:
Generally speaking call to `sched_maybe_preempt` is no-op while interrupts (fixed that) or preemption is disabled.

Procedures `sleep_signal`, `sleepq_broadcast` and so on, may be safely used in interrupt context. I'm quite sure eventually I'll expect callout actions to be executed in thread context, so call to `sched_maybe_preempt` seems ok here.